### ISSUE TITLE
cannot be used to index type (2536) false positive with generic using array length

### ISSUE THREAD

**Issue by DetachHead** [microsoft/TypeScript#45206]
# Bug Report

### 🔎 Search Terms

2536 cannot be used to index type generic

### 🕗 Version & Regression Information

4.4.0-dev.20210727

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.4.0-dev.20210727#code/C4TwDgpgBA+gCgJwpAdgEwHIQB7AwVwFsAeAQSh2AnQGcpSEEBDEY-FAaxQHsB3FAHwCoAXigBtAHTTSAGnriA5ABtqAc2AALRQF0dAKH2hIsAKIoiEZlTIVc1NHQbNW7Ln0HyMdqrSgXCACMrYTFvSgcnJVUUDW0DKESoAH56fSSoAC4zAKsmG3gkVExKAhJSAS8BQwB6ACo69LqoAGMkfIg6Jih2AEtuFChjaG4AM38iYIQ6UYRuQigABiHuKDVqK16WqAADDB2mpqgAARwmQjBVI+GoADFuVbFzS2sIYgBmYRqaxYAfAEZfgAmI7HGgQaCaYDAMA0TLfGjAJgtDjcABuVlGyj4kha8xqTBqADZ3gBOf4ADkWRJBUDqNX0ODA3AQwCG4Ggz0IeRs4XsfgCU1CEyCVh8kSgGHSSVSgqs0sS2RgXJ5b3EOiq4r8AApeihRmLTABKdUKlJQUxm7IoCAYhCGNAQFrKJhIKCjdgtYD9QajB7EAAqWscPU4PH46oE2rRTGU+Ag2QD8j1juwifEKtegei6i0ugEOiN2TR3F6aCAA)

### 💻 Code

```ts
type _PrependNextNum<A extends Array<unknown>> = [...A, A['length']]

type _Enumerate<A extends Array<unknown>, N extends number> = N extends A['length']
    ? A
    : _Enumerate<_PrependNextNum<A>, N>

/**
 * creates a union type of numbers from 0 to generic `N`
 *
 * @example
 * type Foo = Enumerate<3> //0|1|2
 * @see https://stackoverflow.com/a/63918062
 */
export type Enumerate<N extends number> = number extends N
    ? number
    : _Enumerate<[], N> extends (infer E)[]
    ? E
    : never

declare function foo<T extends unknown[]>(
    value: T,
    index: T[Enumerate<T['length']>] //Type 'Enumerate<T["length"]>' cannot be used to index type 'T'.(2536)
    ): void
```

### 🙁 Actual behavior

```
Type 'Enumerate<T["length"]>' cannot be used to index type 'T'.(2536)
```

### 🙂 Expected behavior

no error

**Comment by andrewbranch** [comment]
I feel like this is a super common thing with generic conditional types. You can give the compiler a hint about the constraint of `E`, which will make every branch of `Enumerate` look trivially assignable to `number`. How’s [this](https://www.typescriptlang.org/play?ts=4.4.0-dev.20210727#code/C4TwDgpgBA+gCgJwpAdgEwHIQB7AwVwFsAeAQSh2AnQGcpSEEBDEY-FAaxQHsB3FAHwCoAXigBtAHTTSAGnriA5ABtqAc2AALRQF0dAKH2hIsAKIoiEZlTIVc1NHQbNW7Ln0HyMdqrSgXCACMrYTFvSgcnJVUUDW0DKESoAH56fSSoAC4zAKsmG3gkVExKAhJSAS8BQ2NoGDLghGJw+z8AxtCoDABuQwB6ACoB9IGoAGMkfIg6Jih2AEtuFChaqG4AM38iRrp1hG5CKAAGFe4oNWorebGoAAMMW5GRqAABHCZCMFVn1YAxbjOYnMlmsEGIAGZhH0+kcAD4ARlhACZni8aBBoJpgMAwDRMtCaMAmGMONwAG5WdbKPiSMYHPpMPoANnBAE54QAOI5MlFQAZ9fQ4MDcBDAFbgaDAwh5GwtXyOLZBEKiRWNHyRLrpJKpdpWLWJbIwKUysHiHRVdV+AAU9W2VmI8xQ6ysUFMAgAlGb9SlXd7sigIBSEIY0BAxsomEgoOt2GNgItlusAcQACqWhVuHj8M0CK1kpjKfAQbIp+SO0PYEviY2g1PRdRaXQCHTu7Jk7jzNC9IA)?

**Comment by RyanCavanaugh** [comment]
You could also write this as
```ts
export type Enumerate<N extends number> = number & (number extends N
  ...
```
